File: regions/system/library/dart/health/test/health_test.dart
import 'package:test/test.dart';

import 'package:k1s0_health/health.dart';

class AlwaysHealthy implements HealthCheck {
  @override
  String get name => 'always-healthy';

  @override
  Future<void> check() async {}
}

class AlwaysUnhealthy implements HealthCheck {
  @override
  String get name => 'always-unhealthy';

  @override
  Future<void> check() async => throw Exception('down');
}

void main() {
  group('HealthChecker', () {
    test('チェックがない場合にhealthyを返すこと', () async {
      final checker = HealthChecker();
      final resp = await checker.runAll();
      expect(resp.status, equals(HealthStatus.healthy));
      expect(resp.checks, isEmpty);
    });

    test('全チェックが成功した場合にhealthyを返すこと', () async {
      final checker = HealthChecker()..add(AlwaysHealthy());
      final resp = await checker.runAll();
      expect(resp.status, equals(HealthStatus.healthy));
      expect(resp.checks['always-healthy']?.status, equals(HealthStatus.healthy));
    });

    test('いずれかのチェックが失敗した場合にunhealthyを返すこと', () async {
      final checker = HealthChecker()
        ..add(AlwaysHealthy())
        ..add(AlwaysUnhealthy());
      final resp = await checker.runAll();
      expect(resp.status, equals(HealthStatus.unhealthy));
      expect(resp.checks['always-unhealthy']?.status, equals(HealthStatus.unhealthy));
    });

    test('タイムスタンプを含むこと', () async {
      final checker = HealthChecker();
      final resp = await checker.runAll();
      expect(resp.timestamp, isA<DateTime>());
    });
  });

  group('CheckResult', () {
    test('ステータスとメッセージを保持すること', () {
      const result = CheckResult(status: HealthStatus.degraded, message: 'slow');
      expect(result.status, equals(HealthStatus.degraded));
      expect(result.message, equals('slow'));
    });
  });
}

File: regions/system/library/typescript/health/__tests__/health.test.ts
import { describe, it, expect } from 'vitest';
import { HealthChecker } from '../src/index.js';
import type { HealthCheck } from '../src/index.js';

describe('HealthChecker', () => {
  it('チェックがない場合はhealthyを返す', async () => {
    const checker = new HealthChecker();
    const resp = await checker.runAll();
    expect(resp.status).toBe('healthy');
    expect(Object.keys(resp.checks)).toHaveLength(0);
    expect(resp.timestamp).toBeTruthy();
  });

  it('全チェック成功でhealthyを返す', async () => {
    const checker = new HealthChecker();
    const dbCheck: HealthCheck = {
      name: 'database',
      check: async () => {},
    };
    const cacheCheck: HealthCheck = {
      name: 'cache',
      check: async () => {},
    };
    checker.add(dbCheck);
    checker.add(cacheCheck);

    const resp = await checker.runAll();
    expect(resp.status).toBe('healthy');
    expect(resp.checks['database'].status).toBe('healthy');
    expect(resp.checks['cache'].status).toBe('healthy');
  });

  it('チェック失敗でunhealthyを返す', async () => {
    const checker = new HealthChecker();
    const failCheck: HealthCheck = {
      name: 'database',
      check: async () => { throw new Error('connection refused'); },
    };
    checker.add(failCheck);

    const resp = await checker.runAll();
    expect(resp.status).toBe('unhealthy');
    expect(resp.checks['database'].status).toBe('unhealthy');
    expect(resp.checks['database'].message).toBe('connection refused');
  });

  it('timestampがISO文字列であること', async () => {
    const checker = new HealthChecker();
    const resp = await checker.runAll();
    expect(() => new Date(resp.timestamp)).not.toThrow();
  });
});
